File: speed-cpp/src/SPEED.cpp
#include "SPEED.hpp"
#include <algorithm>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <regex>
#include <sstream>
#include <thread>
#include <vector>

namespace SPEED {

SPEED::SPEED(const std::string &proc_name, const ThreadMode &tmode,
             const std::filesystem::path &speed_dir) {
  self_proc_name_ = proc_name;
  tmode_ = tmode;
  speed_dir_ = speed_dir;
  self_speed_dir_ = speed_dir_ / proc_name;

  if (!Utils::directoryExists(speed_dir_)) {
    Utils::createDefaultDir(speed_dir_);
  }

  if (!Utils::directoryExists(self_speed_dir_)) {
    Utils::createDefaultDir(self_speed_dir_);
  }

  if (!Utils::directoryExists(speed_dir_ / "access_registry")) {
    Utils::createAccessRegistryDir(speed_dir_ / "access_registry");
  }
  // std::cout << "[INFO Speed Dir: " << speed_dir_ << "\n";
  access_list_ = std::make_unique<AccessRegistry>(
      speed_dir_ / "access_registry", proc_name);
}

SPEED::SPEED(const std::string &proc_name, const ThreadMode &tmode)
    : SPEED(proc_name, tmode, Utils::getDefaultSPEEDDir()) {}

SPEED::~SPEED() { kill(); }

bool SPEED::setKeyFile(const std::filesystem::path &key_path) {
  if (!Utils::fileExists(key_path)) {
    std::cout << "[ERROR]: Key File Doesn't Exist. Exitting...\n";
    exit(EXIT_FAILURE);
    return false;
  }
  std::lock_guard<std::mutex> lock(key_mutex_);
  key_ = KeyManager::getKeyFromConfigFile(key_path);
  if (!Utils::validateKey(key_)) {
    std::cout << "[ERROR]: Invalid Key\n";
    throw std::runtime_error("Invalid Key\n");
  }
  key_path_ = key_path;
  return true;
}

void SPEED::setCallback(std::function<void(const PMessage &)> cb) {
  std::lock_guard<std::mutex> lock(callback_mutex_);
  callback_ = std::move(cb);
}

bool SPEED::addProcess(const std::string &proc_name) {
  std::lock_guard<std::mutex> lock(access_list_mutex_);

  if (access_list_) {
    if (!access_list_->check_connection(proc_name)) {
      access_list_->addProcessToList(proc_name);
      Message message = MessageUtils::construct_CON_REQ(proc_name);
      message.header.seq_num = seq_number_.load(std::memory_order_relaxed);
      message.header.sender = self_proc_name_;
      message.header.reciever = proc_name;
      std::vector<uint64_t> k(key_.begin(), key_.end());
      if (!MessageUtils::validate_message_sent(message, self_proc_name_,
                                               proc_name)) {
        std::cout << "[ERROR]: Messsage validation! Before. \n";
        MessageUtils::print_message(message);
        return false;
      }
      EncryptionManager::Encrypt(message, k);
      BinaryManager::writeBinary(message, speed_dir_, seq_number_, proc_name);
      con_res_buffer.insert(proc_name);
      seq_number_.fetch_add(1, std::memory_order_relaxed);
    }
  }

  return true;
}

void SPEED::start() {
  watcher_should_exit_.store(false);

  if (watcher_running_.load())
    return;

  watcher_running_.store(true);

  if (tmode_ == ThreadMode::Single) {
    // In single-thread mode, run watcher in main loop (blocking for
    // bare-metal/embedded)
    watcherSingleThread_();
  } else {
    // Multi-thread mode: watcher runs in background thread
    watcher_thread_ = std::thread([this]() { watcherMultiThread_(); });
    watcher_thread_.detach();
  }
}

void SPEED::stop() { watcher_should_exit_.store(true); }

void SPEED::resume() {
  if (tmode_ == ThreadMode::Multi && !watcher_running_.load()) {
    start();
  }
}

void SPEED::kill() {
  watcher_should_exit_.store(true);
  watcher_running_.store(false);
  access_list_->removeAccessFile();
  stopAllExecutors();
  const std::unordered_set<std::string> acl = access_list_->getAccessList();
  std::vector<uint64_t> k(key_.begin(), key_.end());
  for (const std::string &entry : acl) {
    std::cout << "[DEBUG]: Broadcasting exit notif to: " << entry << "\n\n";
    Message exit_message = MessageUtils::construct_EXIT_NOTIF(entry);
    exit_message.header.seq_num = seq_number_;
    exit_message.header.sender = self_proc_name_;
    EncryptionManager::Encrypt(exit_message, k);
    BinaryManager::writeBinary(exit_message, speed_dir_, seq_number_, entry);

    seq_number_.fetch_add(1, std::memory_order_relaxed);
  }
}

void SPEED::sendMessage(const std::string &msg,
                        const std::string &reciever_name) {
  if (!access_list_->checkGlobalRegistry(reciever_name)) {
    std::cout << "[WARN] Process: " << reciever_name
              << " not in global registry list" << "\n";
    access_list_->incrementalBuildGlobalRegistry();
    std::cout << "[ERROR]: Cannot send Message to: " << reciever_name
              << " as it is not in Global Registry\n";
    return;
  }
  if (!access_list_->checkAccess(reciever_name)) {
    std::cout << "[WARN] Process: " << reciever_name << " not in access list"
              << "\n";
    std::cout << "[ERROR]: Cannot send Message to: " << reciever_name
              << " as it is not in Access List\n";
    return;
  }
  if (!access_list_->check_connection(reciever_name)) {
    std::cout << "[WARN] Process: " << reciever_name
              << " not in connection list" << "\n";
    std::cout << "[ERROR]: Cannot send Message to: " << reciever_name
              << " as it is not in Connection List\n";
    Message con_req_message = MessageUtils::construct_CON_REQ(msg);
    con_req_message.header.seq_num =
        seq_number_.load(std::memory_order_relaxed);
    con_req_message.header.sender = self_proc_name_;
    con_req_message.header.reciever = reciever_name;
    std::vector<uint64_t> k(key_.begin(), key_.end());
    if (!MessageUtils::validate_message_sent(con_req_message, self_proc_name_,
                                             con_req_message.header.sender)) {
      std::cout << "[ERROR]: Messsage validation! Before. \n";
      MessageUtils::print_message(con_req_message);
      return;
    }
    EncryptionManager::Encrypt(con_req_message, k);
    BinaryManager::writeBinary(con_req_message, speed_dir_, seq_number_,
                               con_req_message.header.sender);
    seq_number_.fetch_add(1, std::memory_order_relaxed);
    std::cout << "[INFO]: Connection Request sent to: " << reciever_name
              << "\n";
  }
  Message message = MessageUtils::construct_MSG(msg);

  // load numeric seq value from atomic
  message.header.seq_num = seq_number_.load(std::memory_order_relaxed);

  message.header.sender = self_proc_name_;
  message.header.reciever = reciever_name;
  std::vector<uint64_t> k(key_.begin(), key_.end());
  if (!MessageUtils::validate_message_sent(message, self_proc_name_,
                                           reciever_name)) {
    std::cout << "[ERROR] Message validation failed! Before." << "\n";
    MessageUtils::print_message(message);
    return;
  }
  EncryptionManager::Encrypt(message, k);
  BinaryManager::writeBinary(message, speed_dir_, seq_number_, reciever_name);

  // advance the sender's local sequence number
  seq_number_.fetch_add(1, std::memory_order_relaxed);
}
void SPEED::runWatcherLoop_() {
  while (!watcher_should_exit_.load()) {
    // Scan all new files in our inbox directory
    for (auto &entry : std::filesystem::directory_iterator(self_speed_dir_)) {
      if (!entry.is_regular_file())
        continue;

      auto info = extractFileInfoFromFilename_(entry.path());
      if (!info.has_value())
        continue;

      const std::string fname = entry.path().filename().string();

      // Deduplicate so we don't enqueue the same file twice
      {
        std::lock_guard<std::mutex> seen_lock(seen_mutex_);
        if (seen_.count(fname))
          continue;
        seen_.insert(fname);
      }

      // Build a task and immediately hand it off to the per-receiver executor.
      // This implements "per-thread-per-receiver": each remote (proc_name)
      // gets its own worker that processes that remote's files sequentially.
      Task t;
      t.path = entry.path();
      t.seq = info->seq; // optional metadata

      enqueueTaskForSender(info->proc_name, std::move(t));
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(100));
  }
}

void SPEED::processFile_(const std::filesystem::path &file_path) {
  Message msg;
  try {
    msg = BinaryManager::readBinary(file_path);
  } catch (const std::exception &e) {
    std::cerr << "[SPEED] Failed to read file " << file_path << ": " << e.what()
              << "\n";
    std::lock_guard<std::mutex> lk(seen_mutex_);
    seen_.erase(file_path.filename().string());
    return;
  }

  std::vector<uint64_t> k(key_.begin(), key_.end());
  try {
    EncryptionManager::Decrypt(msg, k);
  } catch (const std::exception &e) {
    std::cerr << "[SPEED] Decrypt failed for " << file_path << ": " << e.what()
              << "\n";
    std::lock_guard<std::mutex> lk(seen_mutex_);
    seen_.erase(file_path.filename().string());
    return;
  }

  if (!MessageUtils::validate_message_recieved(msg, self_proc_name_)) {
    std::cerr << "[ERROR]: Invalid Message received! Not Processing.\n";
    MessageUtils::print_message(msg);
    std::lock_guard<std::mutex> lk(seen_mutex_);
    seen_.erase(file_path.filename().string());
    return;
  }

  // Handle message types
  switch (msg.header.type) {
  case MessageType::MSG: {
    PMessage mm = MessageUtils::destruct_message(msg);
    std::lock_guard<std::mutex> lock(callback_mutex_);
    callback_(mm);
    break;
  }
  case MessageType::EXIT_NOTIF: {
    std::lock_guard<std::mutex> lk(access_list_mutex_);
    access_list_->removeProcessFromGlobalRegistry(msg.header.sender);
    access_list_->removeProcessFromAccessList(msg.header.sender);
    access_list_->removeProcessFromConnectedList(msg.header.sender);
    std::cout << "[DEBUG]: Received EXIT_NOTIF for: " << msg.header.sender
              << "\n";
    break;
  }
  case MessageType::CON_REQ: {
    std::cout << "\n[INFO]: Recieved Connection Request from: "
              << msg.header.sender << "\n";
    auto accessList = access_list_->getAccessList();
    if (accessList.find(msg.header.sender) != accessList.end()) {
      std::cout << "\n[INFO]: Found sender: " << msg.header.sender
                << " in access list\n";
      Message message = MessageUtils::construct_CON_RES(msg.header.sender);
      message.header.seq_num = seq_number_.load(std::memory_order_relaxed);
      message.header.sender = self_proc_name_;
      message.header.reciever = msg.header.sender;
      std::vector<uint64_t> k(key_.begin(), key_.end());
      if (!MessageUtils::validate_message_sent(message, self_proc_name_,
                                               msg.header.sender)) {
        std::cout << "[ERROR]: Messsage validation! Before. \n";
        MessageUtils::print_message(message);
        break;
      }
      EncryptionManager::Encrypt(message, k);
      BinaryManager::writeBinary(message, speed_dir_, seq_number_,
                                 msg.header.sender);
      seq_number_.fetch_add(1, std::memory_order_relaxed);
    } else {
      std::cout << "[ERROR]: Did not find sender: " << msg.header.sender
                << " in access list\n";
    }
    break;
  }
  case MessageType::CON_RES: {
    std::lock_guard<std::mutex> lk(con_res_buffer_mtx_);
    std::cout << "\n[INFO]: Recieved Connection Response from: "
              << msg.header.sender << "\n";
    const std::string con_response_from = msg.header.sender;
    if (con_res_buffer.find(con_response_from) != con_res_buffer.end()) {
      std::cout << "\n[INFO]: Found Connection Response In Buffer from: "
                << msg.header.sender << "\n";
      con_res_buffer.erase(con_response_from);
      access_list_->connect_to(con_response_from);
    }
    break;
  }
  case MessageType::PING:
    pong(msg.header.sender);
    break;
  case MessageType::PONG: {
    PMessage mm = MessageUtils::destruct_message(msg);
    std::lock_guard<std::mutex> lock(callback_mutex_);
    callback_(mm);
    break;
  }
  default:
    break;
  }

  // Always remove file safely
  std::error_code ec;
  std::filesystem::remove(file_path, ec);
  if (ec)
    std::cerr << "[WARN]: Failed to remove file: " << file_path << " -> "
              << ec.message() << "\n";

  std::lock_guard<std::mutex> lock(seen_mutex_);
  seen_.erase(file_path.filename().string());
}

void SPEED::enqueueTaskForSender(const std::string &sender, Task task) {
  std::unique_lock<std::mutex> maplock(executors_mtx_);
  auto &ptr = executors_[sender];
  if (!ptr)
    ptr = std::make_unique<PerSenderExecutor>();
  PerSenderExecutor *pe = ptr.get();
  maplock.unlock();

  {
    std::unique_lock<std::mutex> lk(pe->m);
    pe->q.push_back(std::move(task));
    pe->cv.notify_one();
  }

  bool need_start = false;
  if (!pe->running.load(std::memory_order_acquire)) {
    bool expected = false;
    if (pe->running.compare_exchange_strong(expected, true))
      need_start = true;
  }

  if (need_start) {
    pe->stop_flag.store(false);
    pe->worker = std::make_unique<std::thread>([this, pe, sender]() {
      while (!pe->stop_flag.load()) {
        Task t;
        {
          std::unique_lock<std::mutex> lk(pe->m);
          pe->cv.wait_for(lk, pe->idle_timeout, [&] {
            return !pe->q.empty() || pe->stop_flag.load();
          });

          if (pe->stop_flag.load())
            break;

          if (pe->q.empty())
            continue; // nothing to process, go back to wait
          t = std::move(pe->q.front());
          pe->q.pop_front();
        }

        try {
          this->processFile_(t.path);
        } catch (const std::exception &e) {
          std::cerr << "[SPEED] Worker error for sender " << sender << ": "
                    << e.what() << "\n";
        }
      }
      pe->running.store(false);
    });

    pe->worker->detach();
  }
}

void SPEED::watcherSingleThread_() {
  runWatcherLoop_(); // Blocking call
}

void SPEED::watcherMultiThread_() {
  runWatcherLoop_(); // Non-blocking because start() already spawns thread
}

std::optional<SPEED::ParsedFileInfo>
SPEED::extractFileInfoFromFilename_(const std::filesystem::path &path) const {
  static const std::regex re(
      R"((\d+)_([A-Za-z0-9_]+)_(\d+)_([A-Za-z0-9-]+)\.ospeed)");
  std::smatch m;
  std::string filename = path.filename().string();
  if (std::regex_match(filename, m, re)) {
    try {
      return ParsedFileInfo{m[2].str(),             // proc_name
                            std::stoll(m[3].str()), // seq
                            path};
    } catch (...) {
      return std::nullopt;
    }
  }
  return std::nullopt;
}

void SPEED::ping(const std::string &reciever_name) { ping_(reciever_name); }
void SPEED::pong(const std::string &reciever_name) { pong_(reciever_name); }
void SPEED::ping_(const std::string &reciever_name) {
  Message ping_message = MessageUtils::construct_PING(reciever_name);
  ping_message.header.seq_num = seq_number_;
  ping_message.header.sender = self_proc_name_;
  std::vector<uint64_t> k(key_.begin(), key_.end());
  EncryptionManager::Encrypt(ping_message, k);
  BinaryManager::writeBinary(ping_message, speed_dir_, seq_number_,
                             reciever_name);
  seq_number_.fetch_add(1, std::memory_order_relaxed);
}
void SPEED::pong_(const std::string &reciever_name) {
  Message pong_message = MessageUtils::construct_PONG(reciever_name);
  pong_message.header.seq_num = seq_number_;
  pong_message.header.sender = self_proc_name_;
  std::cout << "\n[INFO]: Sending a PONG to: " << reciever_name << "\n";
  std::vector<uint64_t> k(key_.begin(), key_.end());
  EncryptionManager::Encrypt(pong_message, k);
  BinaryManager::writeBinary(pong_message, speed_dir_, seq_number_,
                             reciever_name);
  seq_number_.fetch_add(1, std::memory_order_relaxed);
}
void SPEED::registerMethod(const std::string &name, RemoteFunction func) {
  function_registry_[name] = std::move(func);
}
void SPEED::invokeMethod(const std::string &name,
                         const std::vector<std::string> &args) {
  if (function_registry_.count(name)) {
    function_registry_[name](args);
  } else {
    std::cerr << "[ERROR] Method '" << name << "' not found.\n";
  }
}

void SPEED::stopAllExecutors() {
  std::lock_guard<std::mutex> maplock(executors_mtx_);
  for (auto &[sender, exec] : executors_) {
    exec->stop_flag.store(true);
    exec->cv.notify_all();
  }
  executors_.clear();
}

} // namespace SPEED

File: speed-python/speed/speed.py
import threading
import time
import queue
from enum import Enum
from pathlib import Path
from typing import Callable, Dict, List, Set, Optional
from concurrent.futures import ThreadPoolExecutor
import filelock

from .utils import Utils
from .access_registry import AccessRegistry
from .binary_manager import BinaryManager
from .encryption_manager import EncryptionManager
from .key_manager import KeyManager
from .message_utils import Message, PMessage, MessageType, MessageUtils

class ThreadMode(Enum):
    SINGLE = 0
    MULTI = 1

class SPEED:
    """
    SPEED IPC implementation for Python
    Compatible with C++ SPEED implementation
    """
    
    def __init__(self, proc_name: str, thread_mode: ThreadMode = ThreadMode.MULTI, 
                 speed_dir: Optional[Path] = None):
        self.proc_name = proc_name
        self.thread_mode = thread_mode
        self.speed_dir = speed_dir or Utils.get_default_speed_dir()
        self.self_speed_dir = self.speed_dir / proc_name
        
        # Initialize components
        self.access_registry = AccessRegistry(self.speed_dir, proc_name)
        self.key = ""
        self.key_path = None
        self.seq_number = 0
        
        # Callback and state management
        self.callback = None
        self.function_registry: Dict[str, Callable] = {}
        self.running = False
        self.watcher_thread = None
        
        # Threading
        self.mutex = threading.RLock()
        self.condition = threading.Condition(self.mutex)
        self.task_queue = queue.Queue()
        self.worker_thread = None
        
        # Create directories
        self._initialize_directories()
        
        # Start the system
        self.start()
    
    def _initialize_directories(self):
        """Initialize required directories"""
        if self.self_speed_dir.exists():
            # Clean up existing directory
            import shutil
            shutil.rmtree(self.self_speed_dir)
        
        Utils.create_default_dir(self.self_speed_dir)
        self.access_registry.sync_access_registry()
    
    def set_key_file(self, key_path: Path) -> bool:
        """Set encryption key from file"""
        try:
            self.key = KeyManager.get_key_from_config_file(key_path)
            self.key_path = key_path
            return Utils.validate_key(self.key)
        except Exception as e:
            print(f"Error setting key file: {e}")
            return False
    
    def set_callback(self, callback: Callable[[PMessage], None]):
        """Set message callback function"""
        self.callback = callback
    
    def send_message(self, message: str, receiver: str):
        """Send message to receiver"""
        with self.mutex:
            if not self.access_registry.check_connection(receiver):
                # Try to establish connection
                if not self._establish_connection(receiver):
                    print(f"Cannot send to {receiver}: not connected")
                    return
            
            # Construct and send message
            msg = MessageUtils.construct_msg(message, self.proc_name, receiver, self.seq_number)
            self.seq_number += 1
            
            # Encrypt if key is set
            if self.key:
                msg = EncryptionManager.encrypt(msg, self.key.encode('utf-8'))
            
            # Write to receiver's directory
            receiver_dir = self.speed_dir / receiver
            filename = f"{msg.header.seq_num:04d}_{Utils.generate_uuid()}.ospeed"
            file_path = receiver_dir / filename
            
            if BinaryManager.write_binary(msg, file_path, self.seq_number, self.proc_name):
                print(f"Message sent to {receiver}")
            else:
                print(f"Failed to send message to {receiver}")
    
    def add_process(self, proc_name: str) -> bool:
        """Add process to access list"""
        with self.mutex:
            self.access_registry.add_process_to_list(proc_name)
            return self._establish_connection(proc_name)
    
    def register_method(self, method_name: str, method: Callable):
        """Register a method for remote invocation"""
        with self.mutex:
            self.function_registry[method_name] = method
    
    def invoke_method(self, method_name: str, receiver: str, args: List[str] = None):
        """Invoke remote method on receiver"""
        with self.mutex:
            if not self.access_registry.check_connection(receiver):
                if not self._establish_connection(receiver):
                    print(f"Cannot invoke method on {receiver}: not connected")
                    return
            
            msg = MessageUtils.construct_invoke_method(method_name, self.proc_name, receiver, args)
            
            # Encrypt if key is set
            if self.key:
                msg = EncryptionManager.encrypt(msg, self.key.encode('utf-8'))
            
            # Send invocation message
            receiver_dir = self.speed_dir / receiver
            filename = f"{msg.header.seq_num:04d}_{Utils.generate_uuid()}.ospeed"
            file_path = receiver_dir / filename
            
            BinaryManager.write_binary(msg, file_path, self.seq_number, self.proc_name)
            self.seq_number += 1
    
    def ping(self, receiver: str):
        """Send ping to receiver"""
        self._send_control_message(receiver, MessageType.PING)
    
    def pong(self, receiver: str):
        """Send pong to receiver"""
        self._send_control_message(receiver, MessageType.PONG)
    
    def _send_control_message(self, receiver: str, msg_type: MessageType):
        """Send control message (PING/PONG)"""
        with self.mutex:
            if not self.access_registry.check_connection(receiver):
                return
            
            if msg_type == MessageType.PING:
                msg = MessageUtils.construct_ping(self.proc_name, receiver)
            else:
                msg = MessageUtils.construct_pong(self.proc_name, receiver)
            
            if self.key:
                msg = EncryptionManager.encrypt(msg, self.key.encode('utf-8'))
            
            receiver_dir = self.speed_dir / receiver
            filename = f"{msg.header.seq_num:04d}_{Utils.generate_uuid()}.ospeed"
            file_path = receiver_dir / filename
            
            BinaryManager.write_binary(msg, file_path, self.seq_number, self.proc_name)
            self.seq_number += 1
    
    def _establish_connection(self, proc_name: str) -> bool:
        """Establish connection with another process"""
        with self.mutex:
            # Check if already connected
            if self.access_registry.check_connection(proc_name):
                return True
            
            # Check if process exists in global registry
            if not self.access_registry.check_global_registry(proc_name):
                # Update registry and check again
                self.access_registry.incremental_build_global_registry()
                if not self.access_registry.check_global_registry(proc_name):
                    print(f"Process {proc_name} not found in global registry")
                    return False
            
            # Send connection request
            con_req = MessageUtils.construct_con_req(self.proc_name, proc_name)
            if self.key:
                con_req = EncryptionManager.encrypt(con_req, self.key.encode('utf-8'))
            
            receiver_dir = self.speed_dir / proc_name
            filename = f"{con_req.header.seq_num:04d}_{Utils.generate_uuid()}.ospeed"
            file_path = receiver_dir / filename
            
            if BinaryManager.write_binary(con_req, file_path, self.seq_number, self.proc_name):
                self.seq_number += 1
                # Wait for connection response
                return self._wait_for_connection_response(proc_name)
            return False
    
    def _wait_for_connection_response(self, proc_name: str, timeout: float = 5.0) -> bool:
        """Wait for connection response"""
        start_time = time.time()
        while time.time() - start_time < timeout:
            if self.access_registry.check_connection(proc_name):
                return True
            time.sleep(0.1)
        return False
    
    def start(self):
        """Start the SPEED system"""
        with self.mutex:
            if self.running:
                return
            
            self.running = True
            
            # Start watcher thread based on mode
            if self.thread_mode == ThreadMode.SINGLE:
                self.watcher_thread = threading.Thread(target=self._watcher_single_thread, daemon=True)
            else:
                self.watcher_thread = threading.Thread(target=self._watcher_multi_thread, daemon=True)
            
            self.watcher_thread.start()
            
            # Start worker thread for processing
            self.worker_thread = threading.Thread(target=self._worker_loop, daemon=True)
            self.worker_thread.start()
    
    def stop(self):
        """Stop the SPEED system"""
        with self.mutex:
            self.running = False
            self.condition.notify_all()
        
        # Send exit notification to connected processes
        self._send_exit_notifications()
        
        # Clean up access file
        self.access_registry.remove_access_file()
    
    def _watcher_single_thread(self):
        """Single-threaded watcher implementation"""
        while self.running:
            self._run_watcher_loop()
            time.sleep(0.1)
    
    def _watcher_multi_thread(self):
        """Multi-threaded watcher implementation"""
        with ThreadPoolExecutor(max_workers=4) as executor:
            while self.running:
                self._run_watcher_loop()
                time.sleep(0.05)
    
    def _run_watcher_loop(self):
        """Core watcher loop"""
        try:
            # Check for new files in our directory
            for file_path in self.self_speed_dir.glob("*.ospeed"):
                self._process_file(file_path)
        except Exception as e:
            print(f"Watcher error: {e}")
    
    def _process_file(self, file_path: Path):
        """Process incoming message file"""
        try:
            # Use file lock to prevent concurrent processing
            lock_file = file_path.with_suffix('.lock')
            with filelock.FileLock(lock_file):
                # Read and parse message
                message = BinaryManager.read_binary(file_path)
                if not message:
                    return
                
                # Decrypt if key is set
                if self.key:
                    message = EncryptionManager.decrypt(message, self.key.encode('utf-8'))
                
                # Validate message
                if not MessageUtils.validate_message_received(message, self.proc_name):
                    file_path.unlink()
                    return
                
                # Process based on message type
                self._handle_message(message)
                
                # Clean up processed file
                file_path.unlink()
                
        except Exception as e:
            print(f"Error processing file {file_path}: {e}")
    
    def _handle_message(self, message: Message):
        """Handle incoming message based on type"""
        if message.header.type == MessageType.MSG:
            self._handle_text_message(message)
        elif message.header.type == MessageType.CON_REQ:
            self._handle_connection_request(message)
        elif message.header.type == MessageType.CON_RES:
            self._handle_connection_response(message)
        elif message.header.type == MessageType.INVOKE_METHOD:
            self._handle_method_invocation(message)
        elif message.header.type == MessageType.PING:
            self._handle_ping(message)
        elif message.header.type == MessageType.PONG:
            self._handle_pong(message)
        elif message.header.type == MessageType.EXIT_NOTIF:
            self._handle_exit_notification(message)
    
    def _handle_text_message(self, message: Message):
        """Handle text message"""
        print(f"🔍 [DEBUG] _handle_text_message called")
        print(f"🔍 [DEBUG] Message type: {message.header.type}")
        print(f"🔍 [DEBUG] Payload length: {len(message.payload)}")
        print(f"🔍 [DEBUG] Payload bytes: {message.payload[:50]}...")  # First 50 bytes
    
        pmsg = MessageUtils.destruct_message(message)
    
        print(f"🔍 [DEBUG] PMessage created:")
        print(f"🔍 [DEBUG]   Sender: {pmsg.sender_name}")
        print(f"🔍 [DEBUG]   Message: '{pmsg.message}'")
        print(f"🔍 [DEBUG]   Message length: {len(pmsg.message)}")
    
        if self.callback:
            print(f"🔍 [DEBUG] Calling user callback...")
            self.callback(pmsg)
        else:
            print(f"Received from {pmsg.sender_name}: {pmsg.message}")    
    def _handle_connection_request(self, message: Message):
        """Handle connection request"""
        sender = message.header.sender
        
        # Check if sender is in our access list
        if self.access_registry.check_access(sender):
            # Accept connection
            self.access_registry.connect_to(sender)
            
            # Send connection response
            con_res = MessageUtils.construct_con_res(self.proc_name, sender)
            if self.key:
                con_res = EncryptionManager.encrypt(con_res, self.key.encode('utf-8'))
            
            sender_dir = self.speed_dir / sender
            filename = f"{con_res.header.seq_num:04d}_{Utils.generate_uuid()}.ospeed"
            file_path = sender_dir / filename
            
            BinaryManager.write_binary(con_res, file_path, self.seq_number, self.proc_name)
            self.seq_number += 1
    
    def _handle_connection_response(self, message: Message):
        """Handle connection response"""
        sender = message.header.sender
        self.access_registry.connect_to(sender)
    
    def _handle_method_invocation(self, message: Message):
        """Handle remote method invocation"""
        try:
            payload = bytes(message.payload).decode('utf-8')
            parts = payload.split(':', 1)
            if len(parts) == 2:
                method_name, args_str = parts
                args = args_str.split(',') if args_str else []
                
                if method_name in self.function_registry:
                    self.function_registry[method_name](args)
                else:
                    print(f"Unknown method: {method_name}")
        except Exception as e:
            print(f"Error invoking method: {e}")
    
    def _handle_ping(self, message: Message):
        """Handle ping message"""
        sender = message.header.sender
        self.pong(sender)
    
    def _handle_pong(self, message: Message):
        """Handle pong message"""
        # Currently just acknowledge
        pass
    
    def _handle_exit_notification(self, message: Message):
        """Handle exit notification"""
        sender = message.header.sender
        self.access_registry.remove_process_from_connected_list(sender)
    
    def _send_exit_notifications(self):
        """Send exit notifications to connected processes"""
        connected = self.access_registry.get_connected_list()
        for proc_name in connected:
            exit_msg = MessageUtils.construct_exit_notif(self.proc_name, proc_name)
            if self.key:
                exit_msg = EncryptionManager.encrypt(exit_msg, self.key.encode('utf-8'))
            
            receiver_dir = self.speed_dir / proc_name
            filename = f"{exit_msg.header.seq_num:04d}_{Utils.generate_uuid()}.ospeed"
            file_path = receiver_dir / filename
            
            BinaryManager.write_binary(exit_msg, file_path, self.seq_number, self.proc_name)
            self.seq_number += 1
    
    def _worker_loop(self):
        """Worker loop for processing tasks"""
        while self.running:
            try:
                # Process tasks from queue if needed
                time.sleep(0.1)
            except Exception as e:
                print(f"Worker error: {e}")
    
    def print_global_registry(self):
        """Print global registry"""
        registry = self.access_registry.get_global_registry()
        print("----------Global Registry----------")
        for proc in sorted(registry):
            print(proc)
        print("----------Global Registry----------")
    
    def print_access_list(self):
        """Print access list"""
        access_list = self.access_registry.get_access_list()
        print("----------Access List----------")
        for proc in sorted(access_list):
            print(proc)
        print("----------Access List----------")
    
    def print_connected_list(self):
        """Print connected list"""
        connected_list = self.access_registry.get_connected_list()
        print("----------Connected List----------")
        for proc in sorted(connected_list):
            print(proc)
        print("----------Connected List----------")
    
    def __enter__(self):
        """Context manager entry"""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.stop()